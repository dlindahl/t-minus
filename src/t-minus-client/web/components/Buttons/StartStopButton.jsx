import { bindActionCreators } from 'redux';
import AppNavButton from './AppNavButton';
import { Component,  PropTypes } from 'react';
import { connect } from 'react-redux';
import Pause from 'react-icons/lib/md/pause';
import PlayArrow from 'react-icons/lib/md/play-arrow';
import { toggleClock } from '../../../shared/actions/ClockActions';

function getState(state) {
  return {
    running: state.clock.running,
    started: !!state.clock.startedAt
  };
}

function getActions(dispatch) {
  return bindActionCreators({ toggleClock }, dispatch);
}

@connect(getState, getActions)
export default class StartStopButton extends Component {
  constructor(props) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }
  handleClick() {
    this.props.toggleClock();
  }
  render() {
    let label;
    if(!this.props.started) {
      label = <PlayArrow data-test-id="PlayArrow"/>;
    } else if(this.props.running) {
      label = <Pause data-test-id="Pause"/>;
    } else {
      label = <PlayArrow data-test-id="PlayArrow"/>;
    }
    return (
      <AppNavButton active={this.props.running} onClick={this.handleClick}>
        {label}
      </AppNavButton>
    );
  }
}

StartStopButton.propTypes = {
  running: PropTypes.bool,
  started: PropTypes.bool
};
StartStopButton.defaultProps = {
  running: false,
  started: false
};
