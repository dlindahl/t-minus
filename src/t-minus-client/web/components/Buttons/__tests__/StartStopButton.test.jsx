import configureStore from 'redux-mock-store'
import emptyObj from 'empty/object'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import StartStopButton from '../StartStopButton'

const STATES = {
  default: {
    clock: emptyObj,
    display: emptyObj
  },
  running: {
    clock: {
      running: true,
      startedAt: Date.now()
    },
    display: emptyObj
  },
  started: {
    clock: {
      startedAt: Date.now()
    },
    display: emptyObj
  }
}
const store = configureStore()

describe('StartStopButton', () => {
  it('renders the component', () => {
    const component = renderer.create(
      <Provider store={store(STATES.default)}>
        <StartStopButton/>
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('is active when the clock that is running', () => {
    const component = renderer.create(
      <Provider store={store(STATES.running)}>
        <StartStopButton/>
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders a label that reflects a clock that has started', () => {
    const component = renderer.create(
      <Provider store={store(STATES.started)}>
        <StartStopButton/>
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders a label that reflects a clock that is running', () => {
    const component = renderer.create(
      <Provider store={store(STATES.running)}>
        <StartStopButton/>
      </Provider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('toggles the clock when clicked', () => {
    const toggleClock = jest.fn()
    const component = renderer.create(
      <Provider store={store(STATES.default)}>
        <StartStopButton.WrappedComponent toggleClock={toggleClock}/>
      </Provider>
    )
    const tree = component.toJSON()
    tree.props.onClick()
    expect(toggleClock).toHaveBeenCalled()
  })
})
