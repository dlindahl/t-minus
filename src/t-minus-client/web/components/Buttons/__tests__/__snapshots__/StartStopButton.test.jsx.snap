exports[`StartStopButton is active when the clock that is running 1`] = `
<button
  disabled={false}
  onClick={[Function]}
  style={
    Object {
      "background": "#34495e",
      "border": 0,
      "color": "#ecf0f1",
      "cursor": "pointer",
      "fontSize": "2rem",
      "height": 50,
      "lineHeight": 1,
      "padding": 0,
      "width": 50,
    }
  }>
  <svg
    data-test-id="Pause"
    fill="currentColor"
    height="1em"
    preserveAspectRatio="xMidYMid meet"
    style={
      Object {
        "color": undefined,
        "verticalAlign": "top",
      }
    }
    viewBox="0 0 40 40"
    width="1em">
    <g>
      <path
        d="m23.4 8.4h6.6v23.2h-6.6v-23.2z m-13.4 23.2v-23.2h6.6v23.2h-6.6z" />
    </g>
  </svg>
</button>
`;

exports[`StartStopButton renders a label that reflects a clock that has started 1`] = `
<button
  disabled={false}
  onClick={[Function]}
  style={
    Object {
      "background": "transparent",
      "border": 0,
      "color": "#bdc3c7",
      "cursor": "pointer",
      "fontSize": "2rem",
      "height": 50,
      "lineHeight": 1,
      "padding": 0,
      "width": 50,
    }
  }>
  <svg
    data-test-id="PlayArrow"
    fill="currentColor"
    height="1em"
    preserveAspectRatio="xMidYMid meet"
    style={
      Object {
        "color": undefined,
        "verticalAlign": "top",
      }
    }
    viewBox="0 0 40 40"
    width="1em">
    <g>
      <path
        d="m13.4 8.4l18.2 11.6-18.2 11.6v-23.2z" />
    </g>
  </svg>
</button>
`;

exports[`StartStopButton renders a label that reflects a clock that is running 1`] = `
<button
  disabled={false}
  onClick={[Function]}
  style={
    Object {
      "background": "#34495e",
      "border": 0,
      "color": "#ecf0f1",
      "cursor": "pointer",
      "fontSize": "2rem",
      "height": 50,
      "lineHeight": 1,
      "padding": 0,
      "width": 50,
    }
  }>
  <svg
    data-test-id="Pause"
    fill="currentColor"
    height="1em"
    preserveAspectRatio="xMidYMid meet"
    style={
      Object {
        "color": undefined,
        "verticalAlign": "top",
      }
    }
    viewBox="0 0 40 40"
    width="1em">
    <g>
      <path
        d="m23.4 8.4h6.6v23.2h-6.6v-23.2z m-13.4 23.2v-23.2h6.6v23.2h-6.6z" />
    </g>
  </svg>
</button>
`;

exports[`StartStopButton renders the component 1`] = `
<button
  disabled={false}
  onClick={[Function]}
  style={
    Object {
      "background": "transparent",
      "border": 0,
      "color": "#bdc3c7",
      "cursor": "pointer",
      "fontSize": "2rem",
      "height": 50,
      "lineHeight": 1,
      "padding": 0,
      "width": 50,
    }
  }>
  <svg
    data-test-id="PlayArrow"
    fill="currentColor"
    height="1em"
    preserveAspectRatio="xMidYMid meet"
    style={
      Object {
        "color": undefined,
        "verticalAlign": "top",
      }
    }
    viewBox="0 0 40 40"
    width="1em">
    <g>
      <path
        d="m13.4 8.4l18.2 11.6-18.2 11.6v-23.2z" />
    </g>
  </svg>
</button>
`;
